### PR TITLE
Fix MacOS YUV Frame Padding

### PR DESCRIPTION
**Pull request by HITGIF**
- Fixes YUV frame padding and alignment
- Uses the pixel resolution of the current display mode (2x)

Performance is still not great, at 3360x2100 CPU utilization is ~200% with ~15fps when things are moving.

### CODE REVIEW

File: src/encoder/ffmpeg.rs [hunk 1]
```diff
@@ -86,12 +88,20 @@ impl FfmpegEncoder {
...
-                    .data_mut()
-                    .copy_from_slice(&nv12[0..self.w * self.h]);
-                frame.planes_mut()[1]
-                    .data_mut()
-                    .copy_from_slice(&nv12[self.w * self.h..]);
+                let encoder_buffer_len = frame.planes_mut()[0].data().len();
+                let encoder_line_size = encoder_buffer_len / self.h;
+                let y = &nv12.luminance_bytes;
+                let uv = &nv12.chrominance_bytes;
+                for (r, row) in enumerate(y.chunks(nv12.luminance_stride as usize)) {
```
Comment (harrynull) [inline src/encoder/ffmpeg.rs:95]: This loop might not always be necessary if there is no padding in the encoder's buffer (i.e. `encoder_line_size == self.w`), and in that case a simple copy might be slightly faster.

File: src/encoder/ffmpeg.rs [hunk 2]
```diff
@@ -86,12 +88,20 @@ impl FfmpegEncoder {
...
+                        .data_mut()[r * encoder_line_size..r * encoder_line_size + self.w]
+                        .copy_from_slice(&row[..self.w])
+                }
+                for (r, row) in enumerate(uv.chunks(nv12.chrominance_stride as usize)) {
+                    frame.planes_mut()[1]
+                        .data_mut()[r * encoder_line_size..r * encoder_line_size + self.w]
+                        .copy_from_slice(&row[..self.w])
+                }
             }
             FrameData::BGR0(bgr0) => match self.pixel_format.as_str() {
```
Comment (harrynull) [inline src/encoder/ffmpeg.rs:106]: This probably also needs a fix for encoder padding. Can be a follow-up.

File: src/encoder/ffmpeg.rs [hunk 3]
```diff
@@ -86,12 +88,20 @@ impl FfmpegEncoder {
...
+                let encoder_buffer_len = frame.planes_mut()[0].data().len();
+                let encoder_line_size = encoder_buffer_len / self.h;
+                let y = &nv12.luminance_bytes;
+                let uv = &nv12.chrominance_bytes;
+                for (r, row) in enumerate(y.chunks(nv12.luminance_stride as usize)) {
+                    frame.planes_mut()[0]
+                        .data_mut()[r * encoder_line_size..r * encoder_line_size + self.w]
+                        .copy_from_slice(&row[..self.w])
+                }
+                for (r, row) in enumerate(uv.chunks(nv12.chrominance_stride as usize)) {
```
Comment (harrynull) [inline src/encoder/ffmpeg.rs:100]: duplicate code? probably can be moved into a helper method.

**Review by harrynull — APPROVED** [review]
mostly looks good.
can you also add a line in README to attribute MirrorX? They're licensed under GPLv3.

File: src/encoder/ffmpeg.rs (hunk 3 above)
Comment (HITGIF) [reply src/encoder/ffmpeg.rs:100]: I think the logic is simple enough to not warrant a helper

**Pull request merged**